### ISSUE TITLE
NativeMidi: music volume change affects master application volume with using of "Microsoft GS" on Windows

### ISSUE THREAD

**Issue by Wohlstand** [WohlSoft/SDL-Mixer-X#7]
It's the weird and clunky imperfection of Microsoft GS Synthesizer which causes application master volume change while trying to change volume of MIDI stream itself. This clunky bug is reproducible on Windows 7, on Windows XP is fine. On Windows 8/8.1/10 is not checked yet.

**The workaround to fix this thing:** inject an extra CC events sequence into raw MIDI stream for all MIDI channels on every volume change request to control MIDI channels volume. The input volume value will be a multiplier that will affect current volume level of each channel to proportionally modify them. This will avoid a usage of master MIDI volume function that causes so clunky behavior.

**Comment by Wohlstand** [comment]
Recently I have found that there are existing a MIDI drivers are don't affecting master MIDI volume change and keeping playing music with full volume.

**Comment by Wohlstand** [comment]
Recently I have re-tested this on Windows 10: this behavior is absolutely same as in the past. Workaround is still be needed.

**Comment by MyreMylar** [comment]
Quick question, as I'm not a MIDI expert and related to the above linked pygame bug.

Would changing the MIDI back-end to something open source like FluidSynth also resolve this issue for a user of SDL Mixer? Assuming those changes of yours eventually make it into the official SDL Mixer version?

**Comment by Wohlstand** [comment]
@MyreMylar, yes, this issue does NOT affects other synthesizers like FluidSynth, Timididty and these are added to MixerX exclusively: libADLMIDI and libOPNMIDI. This issue affects only the NativeMidi, happens only on Windows, happens only in a condition you will use Microsoft GS synthesizer as default MIDI synthesizer. Another way to avoid this issue is installing of a custom system-wide MIDI synthesizer (like Yamaha SyXG50).

**Comment by Wohlstand** [comment]
Because of this issue, a libADLMIDI is used in MixerX as a default system independent MIDI synthesizer.

**Comment by MyreMylar** [comment]
Thanks! That was what I was assuming.

So it sounds like the issue with windows MIDI volume affecting all sounds in pygame might resolve itself if your changes all get ported into the Official SDL Mixer that pygame uses (assuming pygame also updates to the latest SDL at that point).

**Comment by Wohlstand** [comment]
This issue happens with MSGS only and doesn't depends on Mixer itself. Therefore I told that I should to develop a workaround to completely resolve this problem and allow using of Native MIDI safely.

**Comment by rversteegen** [comment]
> Recently I have found that there are existing a MIDI drivers are don't affecting master MIDI volume change and keeping playing music with full volume. Roland Virtual Sound Canvas and Yamaha XG drivers on Windows XP are does that.

I'm not sure I understand what you wrote here. Do you mean that those MIDI drivers are broken and ignore the master MIDI volume? So you think that injecting per-channel volume commands could also fix that problem too?

**Comment by Wohlstand** [comment]
> Do you mean that those MIDI drivers are broken and ignore the master MIDI volume?

**Microsoft GS Wavetable Synthesizer** is only buggy. Yes, per-channel volume injecting should resolve this issue. Look at algorithm:
- keep the factor 1.0...0.0 as a volume level (full or silent)
- on a moment of volume change, multiply last sent CC7 value multiplied to the factor
- every time CC7 is sending by MIDI stream, you should multiply it with to a factor

**Comment by Wohlstand** [comment]
Anyway, current code doesn't allows to implement this workaround quickly, therefore I going to implement some sort of this to use in MixerX instead of old code: https://github.com/Wohlstand/BW_Midi_Sequencer/tree/master/examples/win32-midiOut
However, it's C++, and may looks unapproveable if the pure-C only is needed.

**Comment by Wohlstand** [comment]
Okay, since ages, I desired to finally fix out this. So, meet the alternative NativeMIDI for Windows: https://github.com/WohlSoft/SDL-Mixer-X/commit/8277fe5e912cb252300188561e4f4f9f6be73bca
- I had to use my own MIDI sequencer to extend functionality: full seekability, loop points, XMI/MUS formats support.
- I tricked over 7'th controller and I applied the volume factor to all channels. In result, the workaround for MSGS now works.
